src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentItemDisplayCoordinator.cs

```diff
@@ -226,21 +226,95 @@ public async Task BuildEditorAsync(ContentItem contentItem, BuildEditorContext c
                 var activator = _contentPartFactory.GetTypeActivator(partTypeName);
-                var part = (ContentPart)contentItem.Get(activator.Type, typePartDefinition.Name) ?? activator.CreateInstance();
+                var part = (ContentPart)contentItem.Get(activator.Type, partName) ?? activator.CreateInstance();
+                var isNamedPart = typePartDefinition.PartDefinition.IsReusable() && partName != partTypeName;
+                var partPosition = typePartDefinition.GetSettings<ContentTypePartSettings>().Position ?? "before";
                 part.ContentItem = contentItem;
 
                 // Create a custom shape to render all the part shapes into it
-                var typePartShape = await context.ShapeFactory.CreateAsync("ContentPart_Edit");
+                var shapeType = "ContentPart_Edit";
+                var typePartShapeResult = new ShapeResult(shapeType, ctx => ctx.ShapeFactory.CreateAsync(shapeType));
+                typePartShapeResult.Differentiator($"{contentType}-{partName}");
+                typePartShapeResult.Name(partName);
+                typePartShapeResult.Location($"Parts:{partPosition}");
+
+                typePartShapeResult.Displaying(ctx =>
+                {
+                    // ContentPart_Edit__[PartType]
+                    // eg ContentPart.ServicePart.Edit
+                    ctx.Shape.Metadata.Alternates.Add($"{shapeType}__{partTypeName}");
+
+                    // ContentPart_Edit_[ContentType]_[PartType]
+                    // e.g. ContentPart-LandingPage-ServicePart.Edit
+                    ctx.Shape.Metadata.Alternates.Add($"{shapeType}__{contentType}__{partTypeName}");
+
+                    if (isNamedPart)
+                    {
+                        // ContentPart_Edit__[ContentType]__[PartName]
+                        // e.g. ContentPart-LandingPage-BillingService.Edit ContentPart-LandingPage-HelplineService.Edit
+                        ctx.Shape.Metadata.Alternates.Add($"{shapeType}__{contentType}__{partName}");
+                    }
+                });
+
+                await typePartShapeResult.ApplyAsync(context);
+                var typePartShape = typePartShapeResult.Shape;
+
+                if (typePartShape == null)
+                {
+                    // Part is explicitly noop in placement then stop rendering execution
+                    continue;
+                }
+
                 typePartShape.Properties["ContentPart"] = part;
                 typePartShape.Properties["ContentTypePartDefinition"] = typePartDefinition;
+                partsShape.Properties[partName] = typePartShape;
 
-                var partPosition = typePartDefinition.GetSettings<ContentTypePartSettings>().Position ?? "before";
+                if (isNamedPart)
+                {
+                    // Header zone for named part's heading
+                    typePartShape.Properties["Header"] = await context.ShapeFactory.CreateAsync("Zone");
```

Comment:
I have a bad feeling about this. It's like we have some complex logic existing but here we just say we don't care and inject something because we can.

Reply:
I added it to explicitly render it before all other Field shapes. I guess I can try eliminating this extra zone with `:before` or just not introduce `ContentPartShape` at all

Reply:
Do it

Reply:
Done